ergotree-ir/src/mir/coll_slice.rs

```diff
@@ -0,0 +1,132 @@
+use crate::serialization::op_code::OpCode;
+use crate::serialization::sigma_byte_reader::SigmaByteRead;
+use crate::serialization::sigma_byte_writer::SigmaByteWrite;
+use crate::serialization::SerializationError;
+use crate::serialization::SigmaSerializable;
+use crate::types::stype::SType;
+
+use super::expr::Expr;
+use super::expr::InvalidArgumentError;
+use crate::has_opcode::HasStaticOpCode;
+
+/// Selects an interval of elements
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct Slice {
+    /// Collection
+    pub input: Box<Expr>,
+    /// The lowest index to include from this collection
+    pub from: Box<Expr>,
+    /// The lowest index to exclude from this collection
+    pub until: Box<Expr>,
+}
+
+impl Slice {
+    /// Create new object, returns an error if any of the requirements failed
+    pub fn new(input: Expr, from: Expr, until: Expr) -> Result<Self, InvalidArgumentError> {
+        match input.post_eval_tpe() {
+            SType::SColl(_) => {}
+            _ => {
+                return Err(InvalidArgumentError(format!(
+                    "Expected Slice input to be SColl, got {0:?}",
+                    input.tpe()
+                )))
+            }
+        };
+        if from.post_eval_tpe() != SType::SInt {
+            return Err(InvalidArgumentError(format!(
+                "Slice: expected from type to be SInt, got {0:?}",
+                from
+            )));
+        }
+        if until.post_eval_tpe() != SType::SInt {
+            return Err(InvalidArgumentError(format!(
+                "Slice: expected until type to be SInt, got {0:?}",
+                until
+            )));
+        }
+        Ok(Self {
+            input: input.into(),
+            from: from.into(),
+            until: until.into(),
+        })
+    }
+
+    /// Type
+    pub fn tpe(&self) -> SType {
+        match self.input.post_eval_tpe() {
```

Comment:
The `Slice` result type is the same as `input`, so `self.input.tpe()` should be ok here.

Reply:
Fixed